### PR TITLE
10-70x faster $float, roundtrip + correct rounding via dragonbox

### LINKED ISSUE #104: Fast and accurate float to string conversion (ryu)
I've long been enamored with this one.

https://github.com/ulfjack/ryu

### LINKED ISSUE #13196: Lossy json serialization/deserialization round-trip for floats
A json serialization + deserialization round-trip does not give back the original value for floats.

### Example
```nim
import json
import strformat

# some arbitrary float, not caring about actually significant places here
let x = 0.12345678901234567890123456789
let y = ($(%* x)).parseJson().getFloat()

echo &"{x} == {x} => {x == y}"
doAssert x == y
```
(EDIT(timotheecour): see another example in https://github.com/nim-lang/Nim/issues/15397#issue-707398880)

### Current Output
```
0.1234567890123457 == 0.1234567890123457 => false
/tmp/test.nim(9)         test
/nim/lib/system/assertions.nim(27) failedAssertImpl
/nim/lib/system/assertions.nim(20) raiseAssert
/nim/lib/system/fatal.nim(55) sysFatal
Error: unhandled exception: /tmp/test.nim(9, 10) `x == y`  [AssertionError]
```

### Expected Output

No assertion error at least.

### Possible Solution

This is basically a consequence that `parseFloat` and `$` are not supposed be lossless (see https://github.com/nim-lang/Nim/issues/7717), but the json module builds on these.

I made a quick check with NimYaml and it suffers from exactly the same problem.

It is a natural expectation to rely on `$` to produce an exact string representation. In my opinion the easiest solution would be to improve `$` for floats. Perhaps it would make sense to implement [_Ryū: fast float-to-string conversion_](https://dl.acm.org/doi/10.1145/3192366.3192369) in the Nim standard library to solve the problem for good. A reference implementation is available in [C](https://github.com/ulfjack/ryu) and other languages. This would probably give float-to-string conversion a nice performance boost as well.

```
$ nim -v
Nim Compiler Version 1.1.1 [Linux: amd64]
Compiled at 2020-01-19
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 13ddbc46fc255978ebaf4233009b44db29583cb4
active boot switches: -d:release
```

### LINKED ISSUE #13365: stdlib needs ryu for compact and performant float to string
as of https://github.com/nim-lang/Nim/pull/13364 json works with roundtrip correctness in mind, but sacrifices shortness of representation.
python's json doesnt' have this issue.

### Example
```nim
import std/json
echo(%* 0.6) # 0.59999999999999998
```

### Current Output
```
0.59999999999999998
```

### Expected Output
```
0.6
```

### Possible Solution

port/wrap/implement ryu https://github.com/ulfjack/ryu which has several benefits:
* speed
* compactness

> Ryu generates the shortest decimal representation of a floating point number that maintains round-trip safety. That is, a correct parser can recover the exact original number. For example, consider the binary 32-bit floating point number 00111110100110011001100110011010. The stored value is exactly 0.300000011920928955078125. However, this floating point number is also the closest number to the decimal number 0.3, so that is what Ryu outputs.

## example: python's print wrt float:
as a nice-to-have, we can also have `echo` behave same as in python:
```py
  a = 0.1
  import numpy
  for i in range(0,10):
    print(a)
    a = numpy.nextafter(a, float('Inf'))
```
```
0.1
0.10000000000000002
0.10000000000000003
0.10000000000000005
0.10000000000000006
0.10000000000000007
0.10000000000000009
0.1000000000000001
0.10000000000000012
0.10000000000000013
```

this isn't the case currently in nim
```nim
  var a = 1.0
  for i in 0..<10:
    echo a
    a = nextafter(a, Inf)
```

```
1.0
1.0
1.0
1.000000000000001
1.000000000000001
1.000000000000001
1.000000000000001
1.000000000000002
1.000000000000002
1.000000000000002
```

## links
* https://forum.nim-lang.org/t/5983 Nim's float issue?
* [https://github.com/ulfjack/ryu : Share a paper: How to Print Floating-Point Numbers Accurately - Nim forum](https://forum.nim-lang.org/t/6004)
* [ulfjack/ryu: Converts floating point numbers to decimal strings : ulfjack/ryu: Converts floating point numbers to decimal strings](https://github.com/ulfjack/ryu)
* [disruptek/ryu: ryu for nim : disruptek/ryu: ryu for nim](https://github.com/disruptek/ryu) (archived)
* [Clyybber/nimryu: nim port of ryu : Clyybber/nimryu: nim port of ryu](https://github.com/clyybber/nimryu)
* [LemonBoy/dtoa.nim: Port of Milo Yip's fast dtoa() implementation : LemonBoy/dtoa.nim: Port of Milo Yip's fast dtoa() implementation](https://github.com/LemonBoy/dtoa.nim)
* [Is there a reliable way to round? - Nim forum](https://forum.nim-lang.org/t/7850)
* [Fast and accurate float to string conversion (ryu) · Issue #104 · nim-lang/needed-libraries](https://github.com/nim-lang/needed-libraries/issues/104)
* [⚙ D70631 Microsoft's floating-point to_chars powered by Ryu and Ryu Printf](https://reviews.llvm.org/D70631)
* [alaviss/nim-ryu: An implementation of the ryū float-to-string conversion algorithm. This version is written from scratch based on the paper.](https://github.com/alaviss/nim-ryu)
* [mir-algorithm/generic_128.d at master · libmir/mir-algorithm](https://github.com/libmir/mir-algorithm/blob/master/source/mir/bignum/internal/ryu/generic_128.d)
* [D Ryu implementation reference by 9il · Pull Request #194 · ulfjack/ryu](https://github.com/ulfjack/ryu/pull/194)

## alternative: the newer dragonbox
* [jk-jeon/dragonbox: Reference implementation of Dragonbox in C++](https://github.com/jk-jeon/dragonbox)

### LINKED ISSUE #15397: json: insufficient precision for float64 serialization
Nim's standard json module serializes `float` numbers with 16 significant decimal digits

https://github.com/nim-lang/Nim/blob/fe3211fbcfa6b14a2e2fcd4dfb8664dcf8a1bd21/lib/system/formatfloat.nim#L32

which is insufficient to represent all of 64 bit IEEE-754 floats. One needs at least 17 significant decimal digits, see https://en.wikipedia.org/wiki/Double-precision_floating-point_format for details.

The following MWE shows the problem and a possible solution that changes precision from 16 to 17 digits.

### MWE 
```nim
import json
import fenv
import strformat

let
    x1eps   = 1.0+epsilon(float64) # smallest float64 larger than 1
    jsnStr1 = $(%*{"fnum": x1eps }) # json uses 16 significant digits, not enough!
    jsnStr2 = fmt("""{{ "fnum": {x1eps:.17g} }}""") # need at least 17 significant digits
    jsnNd1  = parseJson(jsnStr1)
    jsnNd2  = parseJson(jsnStr2)
    x1      = jsnNd1{"fnum"}.getFloat(NaN)
    x2      = jsnNd2{"fnum"}.getFloat(NaN)

echo "jsbStr1 = ", jsnStr1
echo "jsbStr2 = ", jsnStr2
echo "jsnNd1  = ", jsnNd1
echo "jsnNd2  = ", jsnNd2
echo "x1      = ", x1
echo "x2      = ", x2

# Results:
doAssert (x1 != x1eps)
doAssert (x1 == 1.0)
doAssert (x2 == x1eps)
```

### Additional Information
```
$ nim -v
Nim Compiler Version 1.2.6 [MacOSX: amd64]
Compiled at 2020-09-02
Copyright (c) 2006-2020 by Andreas Rumpf
git hash: 211868b85cf8b11baabf06d75aaf7e2015d82e2b
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by timotheecour**
## fixes
* closes https://github.com/nim-lang/Nim/issues/13365
* closes https://github.com/nim-lang/needed-libraries/issues/104
* fixes https://github.com/nim-lang/Nim/issues/13196
* fixes https://github.com/nim-lang/Nim/issues/15397
* fixes https://forum.nim-lang.org/t/5983 (see results in https://gist.github.com/timotheecour/e418ea55670b57c325da98a996ff9527)
* fixes https://forum.nim-lang.org/t/7850
```nim
when true: # D20210513T184615
  from math import round
  let a = round(9.779999999999999, 2)
  doAssert a == 9.78
  echo a # before PR: 9.779999999999999, after PR: 9.78, like in python3
``` 
it also makes `round` actually useful for the purpose of formatting

## dragonbox
* replaces `$float`, `addFloat` with dragonbox algorithm
* works in c, cpp, js (rt+vm) [2]
* closes https://github.com/nim-lang/Nim/issues/13365 (dragonbox is faster than ryu [1])
* 10x-70x faster than current $float depending on float distribution, see `nim r -d:danger tests/benchmarks/tstrfloats_bench`
* use `-d:nimLegacyAddFloat` for legacy `addFloat/$`, or use `toStringSprintf` directly
* new low level module `std/strfloats`
* zero impact on CT performance thanks to caching the object file (unless `--forceBuild` is passed)

## new mechanism for managing cacheable dependencies
* new (private, for now) module `std/private/dependency_utils` which allows specifying dependencies at CT, eg:
```nim
import private/dependency_utils
static: addDependency("dragonbox")
  # this will trigger compiler to build a dependency (or use a cached version)
  # and bundle it to the binary we're building, similar to `{.compile: "foo.c".}` but more flexible
```
* this will allow in future work to add more "batteries" to stdlib without having to recompile them, and still allowing them to be used in low-level modules (compiler takes care of installation code which can be arbitrary); in particular for things like bigints (https://github.com/nim-lang/Nim/issues/14696) for which I intend to use the same approach

## design goals
* no compromise on performance, no compromise on compile times
* leverage existing software, don't re-invent the wheel (easily swappable to alternative libraries in future work (or even a pure nim re-implementation), or extensible to other libraries eg bigints)
* transparent change for users (besides the changes in float formatting)

## note
a key part of this PR is that it copies (with minor adjustments) https://github.com/abolz/Drachennest/blob/master/src/dragonbox.cc from this repo: https://github.com/abolz/Drachennest, which uses `Boost Software License 1.0` : https://github.com/abolz/Drachennest/blob/master/LICENSE; that's the only caveat with this PR, but I believe this license should be fine. If not, we could provide compiler options to make this dependency optional (possibly even downloading the sources depending on compiler flags if we must avoid bundling it with nim repo); note that we already have a precedent for that though, with linenoise (bsd2), plus other dependencies past or present.

## links
* [1] see dragonbox vs ryu vs other benchmarks in https://github.com/jk-jeon/dragonbox and https://github.com/abolz/Drachennest
* [2] for js at RT, node/browser already output a minimal representation, before this PR

### CODE REVIEW

**Comment by Araq** [comment]
While I don't mind the "new mechanism for managing cacheable dependencies" mechanism, in fact, I recently had similar ideas, but so that the Nim compiler can use Nimble packages, skimming over dragonbox_impl.cc -- c2nim eats this file for breakfast and then we would have a native Nim implementation. In the past adding C/C++ code didn't work well for us, not just because of the lack of dependency tracking.

**Comment by timotheecour** [comment]
> While I don't mind the "new mechanism for managing cacheable dependencies" mechanism,

great, i think that's a real enabler to add runtime functionality without adding import/CT bloat; 

> in fact, I recently had similar ideas, but so that the Nim compiler can use Nimble packages, 

yes, I'm also thinking in that direction; it's one way to break the cyclic dependency issue

> skimming over dragonbox_impl.cc -- c2nim eats this file for breakfast

how?
```
nimble install c2nim
c2nim -v
0.9.14
c2nim --cpp --debug -o:build/dragonbox_impl_tmp.nim lib/std/private/dragonbox_impl.cc # IndexDefect
```
and from a local clone:
```
$nim_D/c2nim/c2nim --cpp --debug -o:build/dragonbox_impl_tmp.nim lib/std/private/dragonbox_impl.cc

/Users/timothee/git_clone/nim/c2nim/c2nim.nim(192) c2nim
/Users/timothee/git_clone/nim/c2nim/c2nim.nim(141) main
/Users/timothee/git_clone/nim/c2nim/c2nim.nim(76) parse
/Users/timothee/git_clone/nim/c2nim/cparse.nim(2737) parseUnit
/Users/timothee/git_clone/nim/c2nim/cparse.nim(2702) statement
/Users/timothee/git_clone/nim/c2nim/cparse.nim(2636) fullTemplate
/Users/timothee/git_clone/nim/c2nim/cparse.nim(1469) declaration
/Users/timothee/git_clone/nim/c2nim/cparse.nim(2192) compoundStatement
/Users/timothee/git_clone/nim/c2nim/cparse.nim(2712) statement
/Users/timothee/git_clone/nim/c2nim/cparse.nim(2028) declarationOrStatement
/Users/timothee/git_clone/nim/c2nim/cparse.nim(179) backtrackContext
/Users/timothee/git_clone/nim/Nim_devel/lib/system.nim(1888) pop
/Users/timothee/git_clone/nim/Nim_devel/lib/system/fatal.nim(53) sysFatal
Error: unhandled exception: index out of bounds, the container is empty [IndexDefect]
```

> and then we would have a native Nim implementation. In the past adding C/C++ code didn't work well for us, not just because of the lack of dependency tracking.

I'm not against it if there's an automatic source to source translation (that doesn't incur performance loss), but note that even in that case, the `addDependency` would be the way to go IMO, to avoid having to recompile lots of code that should just be compiled once (like in lots of other languages which have a link time dependencies); it's great that nim doesn't need it for most things, but for things like this it's justified, and opens the door for similar future external dependencies.

**Comment by Araq** [comment]
> how?

Well, it needs some c2nim bugfixes but it shouldn't be too hard.

**Comment by planetis-m** [comment]
Why not use the official repo: https://github.com/jk-jeon/dragonbox it has a header only version too. Does it work with float32?

**Comment by timotheecour** [comment]
> Why not use the official repo: jk-jeon/dragonbox it has a header only version too. Does it work with float32?

* the main reason is that it's a bit slower for `float` both in my experience [1] and according to @jk-jeon himself [2]
* a 2nd reason is it requires C++17 instead of C++11 (or perhaps even just C++03, haven't tried with that), but in future c++17 requirement might be relaxed (tracked in https://github.com/jk-jeon/dragonbox/issues/8)
* a minor reason is that jk-jeon/dragonbox is more complex (in part because it has more features)

Note though that jk-jeon/dragonbox supports more options, and we can wrap it in future work to allow customizing (see policy options in jk-jeon/dragonbox, as well as float32 direct support); in fact I've done exactly that, in https://github.com/timotheecour/Nim/pull/732 which wraps both, but we can discuss that after this PR is merged; the good news is that with the logic from this PR, adding new dependencies is easy, more on this later.

[1] see https://github.com/timotheecour/Nim/pull/732 in which I wrap both jk-jeon/dragonbox and abolz/Drachennest; here's my benchmark which shows jk-jeon/dragonbox (aka toStringDragonbox0) is slower for double but faster for float32 compared to abolz/Drachennest (aka toStringDragonbox):
```
("toStringSprintf", "genFloatCast", 11.872696000000001)
("toStringSprintf", "genFloatConf", 1.5924309999999995)
("toStringSprintf", "genFloat32Cast", 5.178619000000001)
("toStringSprintf", "genFloat32Conf", 1.5908620000000013)
("toStringDragonbox", "genFloatCast", 0.16581199999999896)
("toStringDragonbox", "genFloatConf", 0.15630000000000166)
("toStringDragonbox", "genFloat32Cast", 0.22685300000000197)
("toStringDragonbox", "genFloat32Conf", 0.14983400000000202)
("toStringDragonbox0", "genFloatCast", 0.20322699999999827)
("toStringDragonbox0", "genFloatConf", 0.23087699999999955)
("toStringDragonbox0", "genFloat32Cast", 0.14654799999999923)
("toStringDragonbox0", "genFloat32Conf", 0.1509040000000006)
```

[2] see https://github.com/jk-jeon/dragonbox/issues/8#issuecomment-808864341, 
> FYI, @abolz has reimplemented Dragonbox, and AFAIK this implementation does not have lots of compiler-torturing nonsenses mine has, so it is probably C++11 or C++03. He also has an arguably better decimal-to-string conversion routine, which requires a little bit more preconditions but has greater performance. You might be interested to check out the repository.

> Well, it needs some c2nim bugfixes but it shouldn't be too hard.

I honestly doubt it, there'll always be C++ constructs it can't handle; while abolz/Drachennest is C++11 (or c++03), other projects we may want to depend on in future (eg bigint, etc) might have more complex requirements (eg `jk-jeon/dragonbox` uses C++17). Writing a C++ parser that works with real life libraries is not only incredibly difficult, but also entirely un-necessary, cling and clang's libtooling is what should be used for that. But that can be discussed elsewhere.

File: lib/system.nim [hunk 1]
```diff
@@ -112,7 +112,7 @@ proc compileOption*(option, arg: string): bool {.
       discard "compiled with optimization for size and uses Boehm's GC"
 
 {.push warning[GcMem]: off, warning[Uninit]: off.}
-{.push hints: off.}
+# {.push hints: off.}
```
Comment (timotheecour) [inline lib/system.nim]: I could revert the change i did here if needed, but what was the point of this? it prevents useful things, eg honoring `--processing:filenames`, or `--hint:cc` for modules (recursively) imported by system, which I want to see when compiling with those options (instead of obfuscating those). 

`{.push hints: off.}` was also preventing seeing the `CC` hint for the dragonbox sources.

File: tests/benchmarks/tstrfloats_bench.nim [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+nim r -d:danger -d:numIter:100_000_00 tests/benchmarks/tstrfloats_bench.nim
+("toStringSprintf", "genFloatCast", 11.956240000000001)
+("toStringSprintf", "genFloatConf", 1.581176000000001)
+("toStringDragonbox", "genFloatCast", 0.1652149999999999)
+("toStringDragonbox", "genFloatConf", 0.15221700000000027)
+]#
+
+import std/[times, strfloats]
+
+const numIter {.intdefine.} = 10
```
Comment (timotheecour) [inline tests/benchmarks/tstrfloats_bench.nim:12]: so that it runs fast in CI, but users can run a more meaningful benchmark with larger `numIter` (as explained in benchmarks/readme.md)

File: tests/stdlib/tjsonutils.nim [hunk 3]
```diff
@@ -4,15 +4,38 @@ discard """
...
+  # note that this isn't always possible, e.g. for pointer-like types.
+  let j = t.toJson
+  let j2 = $j
+  doAssert j2 == expected, j2
+  let j3 = j2.parseJson
+  let t2 = j3.jsonTo(T)
+  doAssert t2 == t # xxx handle NaN specially, using isNaN once https://github.com/nim-lang/Nim/issues/18007 is fixed
+  doAssert $t2.toJson == j2 # still needed, because -0.0 = 0.0 but their json representation differs
+
+proc testRoundtrip2[T](t: T, expected: string) =
```
Comment (timotheecour) [inline tests/stdlib/tjsonutils.nim]: redundant with `testRoundtripVal`, will remove this in next PR or this PR if there are other things to address

File: tests/stdlib/tstrfloats.nim [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+          toStr(NaN) == "nan"
+          toStr(Inf) == "inf"
+          toStr(1.0 / 0.0) == "inf"
+          toStr(-1.0 / 0.0) == "-inf"
+          toStr(50e18) == "5e+19"
+
+    block:
+      let a = 123456789
+      when not defined(js):
+        # xxx in VM, gives: Error: VM does not support 'cast' from tyInt to tyFloat
```
Comment (timotheecour) [inline tests/stdlib/tstrfloats.nim]: refs https://github.com/timotheecour/Nim/issues/733

File: compiler/builddeps.nim [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  of "dragonbox":
+    if name notin conf.dependencies:
+      conf.dependencies.add name
+      # xxx we could also build this under $nimb/build/
+      let dir = conf.getNimcacheDir().string
+      createDir dir
+      let objFile = dir / ("$1nimdragonbox.o" % prefix)
+      if optForceFullMake in conf.globalOptions or not objFile.fileExists:
+        # xxx
+        # let cppExe = getCompilerExe(c.config; compiler: TSystemCC; cfile: AbsoluteFile): string =
```
Comment (timotheecour) [inline compiler/builddeps.nim]: I'm currently hardcoding the C++ compiler; followup PR will improve this to use `extccomp.getCompilerExe` instead so it honors user configs (deferring this to future PR as it requires a few other changes to allow this and this PR is already quite big)

Comment (Araq) [reply compiler/builddeps.nim]: We should c2nim the code, as I said.

Comment (timotheecour) [reply compiler/builddeps.nim]: as I wrote above, I've tried that already; c2nim can't handle dragonbox.cc, it lacks a lot of functionality to support C++ files like dragonbox.cc and chokes on many different parts of the code (see also https://github.com/nim-lang/c2nim/issues).
Compiling `dragonbox.cc` directly as in this PR is much simpler than fixing c2nim to parse C++ properly and convert it to nim (only to be later converted back to C or C++ during cgen); fixing c2nim could easily represent a multi-year dedicated effort, time that could be better spent elsewhere.

**Comment by planetis-m** [comment]
Anyway if people start converting it to Nim like the last time with ryu, we should probable coordinate and split tasks.

**Comment by abolz** [comment]
Thanks for trying Drachennest! :-) Nice to see that this might actually be useful for someone else.

Just a few remarks:

The Dragonbox implementation in Drachennest currently only works correctly for double-precision (float64) numbers. float32 is actually not supported yet. (The double-precision implementation might produce more digits than necessary for single-precision numbers.)

So you should actually use Junekey Jeon's implementation (which currently requires C++20). Or the simplified implementation of Dragonbox in the fmt formatting library [here](https://github.com/fmtlib/fmt/blob/54f22a3eef27c1e29cf0158945ff50bbab232d27/include/fmt/format-inl.h#L828) (contributed by Junekey Jeon), which works with C++11, produces output similar to printf("%g"), is widely used and well tested.

Alternatively you might want to try out the Schubfach implementation ([float32 version](https://github.com/abolz/Drachennest/blob/e6714a39ad331b4489d0b6aaf3968635bff4eb5e/src/schubfach_32.cc), [float64 version](https://github.com/abolz/Drachennest/blob/e6714a39ad331b4489d0b6aaf3968635bff4eb5e/src/schubfach_64.cc)). Dragonbox is an optimized version of this algorithm. It might be slightly slower than Dragonbox, but the implementation is less complex. (You may tweak the output using the MinFixedDecimalPoint/MaxFixedDecimalPoint constants in the ToChars methods to better match the output of printf.) If you want to try this out, I may try to simplify the code such that (the current version of) c2nim is able to translate it.

HTH

**Comment by timotheecour** [comment]
> Thanks for trying Drachennest!

and thanks for writing this library!
> float32 is actually not supported yet

indeed, as can be seen with: `let a = 1.1'f32`: when converting this to float64 first you'd get `1.100000023841858`, whereas when using `jk-jeon/dragonbox` (eg via https://github.com/timotheecour/Nim/pull/732) you'd get `1.1`

But note that this isn't a regression, because nim currently renders float32's by first converting it to float64 (and then using sprintf, until this PR), so with this PR you still get all the benefits of dragonbox for `float/float64` (arguably more common in nim code than float32), and you get possibly more digits than needed for float32's, but still better (or at least no worse) even in that case compared to before this PR.

Future work can add support for float32's, which will require other changes to nim anyways to avoid the pre-exixting conversions from float32 to float64 done in a few places.
(and will probably use https://github.com/abolz/Drachennest/blob/e6714a39ad331b4489d0b6aaf3968635bff4eb5e/src/schubfach_32.cc for the float32 case, and the code from this PR for the float64 case)

> Anyway if people start converting it to Nim like the last time with ryu, we should probable coordinate and split tasks.

I don't see the point in doing that, you wouldn't get the benefits of future improvements from upstream libraries, and it'd have to be compiled in to object code anyways (as done in this PR) to avoid adding ever more import/cyclic dependencies just to get a working `$float`. If a suitable library written in nim comes along in the future that has no worse performance than the ones used in this PR, we'd always be able to swap it in at that time (as pre-compiled dependency), but until then I'd very much like to benefit from what those libraries offer without having to wait for this hypothetical rewrite with possibly worse performance.

**Comment by timotheecour** [comment]
@araq PTAL

I have another branch (which depends on this PR) where it also applies dragonbox algorithm to float32 (using Drachennest-schubfach_32) in addition to float64 (still using Drachennest-dragonbox), along with performance benchmarks justifying it; it builds on this PR's `addDependency` logic, and gives the fastest performance possible + roundtrip + correct rounding in all scenarios: (c,cpp,js), (rt,ct), (float, float32), and only needs C++11 (possibly C++03) for compilation of the cached dependency (but `nim c` continues to work even though C++11 is used for the dependency).

File: tests/stdlib/tjsonutils.nim (hunk 3 above)
Comment (timotheecour) [reply tests/stdlib/tjsonutils.nim]: done

File: tests/stdlib/tstrfloats.nim (hunk 4 above)
Comment (timotheecour) [reply tests/stdlib/tstrfloats.nim]: incorporated in code comment

File: tests/benchmarks/tstrfloats_bench.nim (hunk 2 above)
Comment (timotheecour) [reply tests/benchmarks/tstrfloats_bench.nim:12]: added a note in tests/benchmarks/readme.md to explain this

File: lib/system.nim (hunk 1 above)
Comment (timotheecour) [reply lib/system.nim]: I've now reverted this diff and moved it in a dedicated PR, see https://github.com/nim-lang/Nim/pull/18034

**Comment by Araq** [comment]
c2nim can now parse dragonbox.cc and the way you did the "new mechanism for managing cacheable dependencies" via yet another compiler magic is not acceptable. We need an external "bundler" tool that is available at boot time (maybe by generating C sources for it).

**Comment by timotheecour** [comment]
> @c2nim can now parse dragonbox.cc

(EDIT with `--keepBodies`) here's the output of `c2nim --cpp --keepBodies --debug -o:output.nim lib/vendor/drachennest/dragonbox.cc` as of latest c2nim, 98af60a3484696ab1b138789abc2f82c2ba733d0:

https://gist.github.com/timotheecour/8f45eddc5b3533f5e166b305fde537f9

definitely better than before.

But right now, c2nim doesn't produce a valid nim file, `nim c output.nim` complains, and in after trying to manually fix ~20 different kinds of errors i stopped because i had no idea how many more errors there would be, nor whether the produced nim file would work in the end.

if c2nim ever becomes capable of generating correct nim code (without performance loss) for dragonbox.cc, that's great, we'll always have the option to replace dragonbox.cc with c2nim_dragonbox.nim at that point, but until then, I don't see anything wrong with using dragonbox.cc directly as I did in this PR, so we can benefit from this PR's speedup and correctness improvements without waiting for c2nim.

> and the way you did the "new mechanism for managing cacheable dependencies" via yet another compiler magic is not acceptable. We need an external "bundler" tool that is available at boot time (maybe by generating C sources for it).

That was my first thought, but doing this at boot time is strictly less flexible, and prevents optional, on-demand dependencies (bigint etc, for which i have a PR in the work using the same approach leveraging existing libraries).

`addDependency("dragonbox")` is analog to `{.compile: "dragonbox.cc".}` and  `{.link: "libdragonbox.so".}`, and allowing it in user code instead of a separate pre-processing step gives you flexibility. In the future, once addDependency API is stabilized with more use cases (bigint etc), we can also expose it to user code in a more granular way, allowing to specify how to build a dependency.

> yet another compiler magic

There are plenty of magics that could be replaced by regular procs (eg `abs`, etc), but this is a case where using a magic makes sense, as it allows using full power of stdlib without causing import dependencies for user code (hence usable in low level code like strmantle).

**Comment by kunitoki** [comment]
I agree with Araq we should port those algos to nim, so there is no need to rush a new dependency injection system in the compiler. It also has the benefit that you can use them both at runtime, compile time and in nimscript, while using the wrapper to dragonbox.cc you get different results in different contexts. 

And i add that we should have both float32 and 64 versions.

**Comment by timotheecour** [comment]
> I agree with Araq we should port those algos to nim

then please write the port and submit a PR after making sure the performance is comparable or better to what this PR offers? I don't see the point in re-inventing the wheel, this is NIH.

>  It also has the benefit that you can use them both at runtime, compile time and in nimscript

so, just like this PR then?

> while using the wrapper to dragonbox.cc you get different results in different contexts.

how so?

> And i add that we should have both float32 and 64 versions.

I have a followup branch that supports float32 as mentioned in https://github.com/nim-lang/Nim/pull/18008#issuecomment-841753138

**Comment by kunitoki** [comment]
I'm not sure, it seems the usage of dragonbox is disabled for nimscript https://github.com/nim-lang/Nim/pull/18008/files#diff-f056834522efcf8e1e24c2c0cb2408cb1b1e6da5ac7a28129f514b89d3134394R17 but maybe it's just me being a newbie in nim compiler code... 

I didn't want to sound harsh, but just wanted to say that having proper nim support could lead to simplified internals and better control on the compiler code.

**Comment by timotheecour** [comment]
> I'm not sure, it seems the usage of dragonbox is disabled for nimscript #18008 (files) but maybe it's just me being a newbie in nim compiler code...

see for yourself: `nim --eval:'var a = 0.1; echo (a+0.2)'` shows 0.30000000000000004 (or try with a nimscript file) after this PR

> simplified internals

nim should be able to handle external dependencies (and already does with pcre and all the existing wrappers); re-implementing existing libraries in nim doesn't reduce the overload workload/maintenance, the opposite is true.

**Comment by kunitoki** [comment]
Then what  `const useDragonbox = ... and not defined(nimscript)` means ?

**Comment by timotheecour** [comment]
> Then what const useDragonbox = ... and not defined(nimscript) means ?

it's a vmops, see `registerCallback c, "stdlib.strfloats.addFloat", proc(a: VmArgs) =`

**Comment by timotheecour** [comment]
closing this because https://github.com/nim-lang/Nim/pull/18139 was merged, but see https://github.com/nim-lang/Nim/pull/18139#issuecomment-852636577 which shows a 1.4X performance drop compared to this PR
